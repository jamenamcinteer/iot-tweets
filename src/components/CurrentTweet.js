import React, { useEffect, useState } from "react"
import moment from "moment"
import NextTweetButton from "./NextTweetButton"

const CurrentTweet = ({ twitterData }) => {
  const [currentTweet, setCurrentTweet] = useState()

  useEffect(() => {
    const newIndex = Math.floor(Math.random(0, 1) * twitterData.length)
    console.log(twitterData[newIndex])
    setCurrentTweet(twitterData[newIndex])
    // setInterval(() => {
    //   const newIndex = Math.floor(Math.random(0, 1) * twitterData.length)
    //   setCurrentTweet(twitterData[newIndex])
    // }, 15000)
  }, [twitterData])

  const nextClickHandler = () => {
    const newIndex = Math.floor(Math.random(0, 1) * twitterData.length)
    setCurrentTweet(twitterData[newIndex])
  }

  // BUG: moment returns a warning regarding the date format that Twitter provides us
  return (
    <>
    {currentTweet && (
      <div>
        <img src={currentTweet.user_profile_image_url_https} alt="" />
        <p>{currentTweet.user_screen_name} @{currentTweet.user_name} <span className="tweet__date">Â· {moment(currentTweet.created_at).fromNow()}</span></p>
        <p dangerouslySetInnerHTML={currentTweet}></p>
        <NextTweetButton clickHandler={nextClickHandler} />
      </div>
    )}
    </>
  )
}

export default CurrentTweet