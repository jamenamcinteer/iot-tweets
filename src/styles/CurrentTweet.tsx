import styled from "styled-components"
import { IPropsTheme } from "../interfaces/interfaces"

export const TweetUserProfileImageContainer = styled.div`
  border-radius: 9999px;
  box-shadow: ${(props: IPropsTheme) => props.theme.boxShadowSecondary};
  display: inline-block;
`

export const TweetUserProfileImage = styled.img`
  border-radius: 9999px;
  display: block;
  /* -webkit-mask-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50"/></svg>'); */
  /* mask-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50"/></svg>'); */
  /* mask-image: url('data%3Aimage%2Fsvg%2Bxml%3Butf8%2C%3Csvg%20viewBox%3D%220%200%20100%20100%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%2F%3E%3C%2Fsvg%3E'); */
  /* mask-mode: alpha; */
  /* -webkit-mask-repeat: no-repeat; */
  /* mask-repeat: no-repeat; */
  /* -webkit-mask-size: contain; */
  /* mask-size: contain; */
  /* -webkit-mask-position: center; */
  /* mask-position: center; */
`

export const TweetDate = styled.span`
  white-space: nowrap;
`

export const TweetMeta = styled.p`
  font-size: ${(props: IPropsTheme) => props.theme.fontSizeSmall};
`

export const TweetText = styled.p`
  font-size: ${(props: IPropsTheme) => props.theme.fontSizeMedium};
`