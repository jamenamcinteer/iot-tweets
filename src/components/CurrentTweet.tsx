import React, { useEffect, useState } from "react"
import moment from "moment"
import NextTweetButton from "./NextTweetButton"
import { ITweets, ICurrentTweetProps } from "../interfaces/interfaces"
import { TweetContainer, TweetUserProfileImageContainer, TweetUserProfileImage, TweetDate, TweetMeta, TweetText } from "../styles/CurrentTweet"

function CurrentTweet({ twitterData }: ICurrentTweetProps) {
  const [currentTweet, setCurrentTweet] = useState<ITweets | undefined>()

  // Select a random tweet to display when the component gets the twitter data
  useEffect(() => {
    const newIndex: number = Math.floor(Math.random() * twitterData.length)
    setCurrentTweet(twitterData[newIndex])
  }, [twitterData])

  // Change the current tweet to another random tweet when the Another Tweet button is clicked
  const nextClickHandler = () => {
    const newIndex: number = Math.floor(Math.random() * twitterData.length)
    setCurrentTweet(twitterData[newIndex])
  }

  return (
    <>
    {currentTweet && (
      <TweetContainer>
        <TweetUserProfileImageContainer>
          <TweetUserProfileImage src={currentTweet.user_profile_image_url_https} alt="" data-testid="tweet-img" />
        </TweetUserProfileImageContainer>
        <TweetMeta data-testid="tweet-meta">{currentTweet.user_screen_name} @{currentTweet.user_name} <TweetDate>Â· {moment(currentTweet.created_at, "ddd MMM D HH:mm:ss ZZ yyyy").fromNow()}</TweetDate></TweetMeta>
        <TweetText data-testid="tweet-text" dangerouslySetInnerHTML={currentTweet}></TweetText>
        <NextTweetButton clickHandler={nextClickHandler} />
      </TweetContainer>
    )}
    </>
  )
}

export default CurrentTweet