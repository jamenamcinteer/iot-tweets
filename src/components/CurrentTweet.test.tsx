import React from 'react';
import { cleanup, render, fireEvent, wait } from "@testing-library/react";
import "jest-styled-components";
import { mockSuccessResponseTwitter } from "../__mocks__/mockTwitterApiResponse"
import CurrentTweet from './CurrentTweet';

afterEach(cleanup);

it('renders as expected', () => {
  const realDateNow = Date.now.bind(global.Date);
  const dateNowStub = jest.fn(() => 1574012671025);
  global.Date.now = dateNowStub;

  const mockMath = Object.create(global.Math);
  mockMath.random = () => 0.01;
  global.Math = mockMath;

  const { asFragment } = render(<CurrentTweet twitterData={mockSuccessResponseTwitter} />)
  expect(asFragment()).toMatchSnapshot()

  global.Date.now = realDateNow;
})

it('should call clickHandler when clicked', () => {
  const { getByText, getByTestId } = render(<CurrentTweet twitterData={mockSuccessResponseTwitter} />)
  const oldImg = getByTestId("tweet-img")
  const oldMeta = getByTestId("tweet-meta")
  const oldText = getByTestId("tweet-text")

  fireEvent.click(getByText('Another Tweet'))

  setTimeout(() => {
    expect(getByTestId("tweet-img")).not.toBe(oldImg)
    expect(getByTestId("tweet-meta")).not.toBe(oldMeta)
    expect(getByTestId("tweet-text")).not.toBe(oldText)
  }, 10)
  
})