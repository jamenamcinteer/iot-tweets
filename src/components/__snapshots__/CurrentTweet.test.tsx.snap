// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<DocumentFragment>
  .c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: transparent;
  border: 0;
  border-radius: 1.6rem;
  padding: 1rem 1.5rem;
  cursor: pointer;
}

.c0 {
  border-radius: 9999px;
  display: inline-block;
}

.c1 {
  border-radius: 9999px;
  display: block;
}

.c2 {
  white-space: nowrap;
}

<div>
    <div
      class="c0"
    >
      <img
        alt=""
        class="c1"
        data-testid="tweet-img"
        src="https://pbs.twimg.com/profile_images/823638186429558787/D8CBNuC3_normal.jpg"
      />
    </div>
    <p
      class=""
      data-testid="tweet-meta"
    >
      spiercy82 @Sean Piercy 
      <span
        class="c2"
      >
        · 19 hours ago
      </span>
    </p>
    <p
      class=""
      data-testid="tweet-text"
    >
      RT @clark_robotics: Adobe has slashed the price of Photoshop in its early Black Friday sale: #ai #ml #dl #iot Cc: @MIKEQUINDAZZI https://t.…
    </p>
    <button
      class="c3"
    >
      Another Tweet
      <svg
        fill="none"
        style="width: 14px; height: 14px; margin-left: 5px;"
        viewBox="0 0 12 12"
      >
        <path
          d="M6 0.666664L5.06 1.60666L8.78 5.33333H0.666668V6.66666H8.78L5.06 10.3933L6 11.3333L11.3333 6L6 0.666664Z"
          fill="white"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
